enhancements/installer/credentials-management-outside-openshift-cluster.md

```diff
@@ -0,0 +1,130 @@
+
+### Risks and Mitigations
+
+## Design Details
+
+We intend to build an optional tool **ccoctl** which will handle credentials management of the cluster in *manual* mode. 
+The following is the set of requirements for the current prototype (for AWS). We can have something similar for other 
+platforms.
+
+* ccoctl should be able to setup Identity Provider to authenticate OpenShift components
+* ccoctl should be able to take a list of CredentialsRequests from the release image and create/update 
+  Roles/Users/ServicePrincipals/ServiceAccounts with appropriate permissions.
+* ccoctl should be able to take list of CredentialsRequests for a release image, and the Identity Provider URL to generate 
+  the objects that need to be passed to the installer for successful installation
+* ccoctl should be able to delete all the resources that it had created in the cloud
+
+We envision **ccoctl** as a recommended tool to setup credentials on new platforms but customers are free to use other 
+tools like Terraform/AWS CloudFormation to do the above mentioned tasks. Read more about AWS implementation details [here](https://github.com/openshift/cloud-credential-operator/blob/master/docs/ccoctl.md)
+
+We also plan to have a detailed documentation in place to guide new cloud providers to implement *manual* mode.
+
+### Open Questions
+
+### Test Plan
+
+We plan to have a e2e test that will externally set up a credentials management infrastructure and then kickstart 
+install in a *manual* mode.
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+To upgrade we need to execute following steps:
+* Examine the CredentialsRequests in the new OpenShift release. Check if permissions in the existing CredentialsRequest
+  have changed.
+* Create/update credentials in the underlying cloud provider, and also create/update Kubernetes Secrets in the correct 
+  namespaces to satisfy all CredentialsRequests in the new release.
+* Set an appropriate annotation `cloudcredential.openshift.io/upgradeable-to` to a new upgradable version.
+
+We also plan to have provision in ccoctl to help in upgrades. Details in [this](https://issues.redhat.com/browse/CCO-84) card.
+
+### Version Skew Strategy
+
+## Implementation History
+
+We currently have a work-in-progress CLI tool [ccoctl](https://github.com/openshift/cloud-credential-operator/blob/master/docs/ccoctl.md) 
+to create and manage cloud credentials outside the cluster for AWS cloud. The design details of this tool is discussed above.
+
+## Drawbacks
+
+* Taking the credentials management outside the cluster will create additional overhead for the customer to make sure all
+  the required infrastructure is in place before starting the installation process. Current tooling we have only supports
+  the AWS cloud, we do not have anything planned for other cloud providers.
+* Push-button upgrades will not work in *manual* mode as the cluster no longer has the admin credentials to mint credentials 
```

Comment:
I'm sure that this has been discussed at length, but this is a huge drawback. Going forward we would be saying that, by default, we do not support push-button upgrades.

Reply:
That part that I am tripping up on is that the second sentence of the summary of this enhancement is "We propose to make *manual* mode as default for clusters on new platforms." That takes away the "opt-in" aspect of Manual mode.

Reply:
That's a good point, this is what we are proposing for a new cloud provider we expect to be added. All clusters on that platform would initially be in this mode. Mint may never happen on that platform, we're not sure.

Many providers today are passthrough only which is technically masking some of the problems we've tried to solve with manual mode, where nothing is actively preventing you from upgrading when your passthrough shared creds are lacking some new permission. So we're not forcing them to take action for upgrades but at the cost of potential breakage. I think these cloud providers tend to use pretty broad permissions so I don't recall seeing it bite anyone yet.

So I guess here, for new platforms, we are proposing that this is now the default, you do not have automated 4.x to 4.x+1 upgrades, you'll need to take action manually to clarify that you've prepared for it.

It should be noted that this does not apply to .z upgrades, 4.7.1 to 4.7.2 is automated without user interaction.

Reply:
I don't know if this helps: the push-button upgrades are available as long as the cluster admin sets the appropriate annotation. Once the annotation is in place, the upgrade button will be available (whether the user has actually completed the necessary work will show up during/post upgrade).

The new requirement for doing some pre-upgrade steps is a direct consequence of the decision to move cloud/platform credentials management out of the cluster. Given this new directive, we're left with something needing to tell the cluster that is okay to perform an upgrade (since it can no longer self-service). We could encode this in the release image so that CCO/CVO can peek into the next version to know about the upgradeability...but I'm just brainstorming here and I don't think we've spent much time considering how to improve this new upgrade experience.

Reply:
can we separate patch release upgrades from minor release upgrades in this flow?

it is a _good thing_ that we are clear with our users on the permissions we require when they differ across releases.

is it possible to integrate whatever checking is required direct in the cvo itself when it evaluates the payload for other characteristics?

Reply:
@LalatenduMohanty @wking WDYT about including this logic in CVO?

Reply:
I understand the need for manual mode from customers. But are we sure we want to make it the default? Why not keeping mint mode defult but manual mode for customer who want this? May be this has been discussed before.  I am asking automatic upgrade was part of OCP/OTA plan and making the manual mode as default is a big change. I hope product management has already signed-off on this. 

I agree that we should separate Y stream updates from Z stream updates. We should keep the automatic updates enabled for Z stream updates if that is possible by changing CVO  to validate if the cluster has right permissions, I am fine with the idea. 

We would like to know more about the checks we need to add to CVO and then we can say more clearly if it will be feasible with respect to CVO's architecture.

Reply:
The idea is that the OpenShift product will get out of the business of self-managing cloud permissions/credentials. Mint mode would not be implemented today based on this directive. And based on this directive, supporting AWS Role-based/Pod Identity with short lived tokens (STS) is done outside of the cluster as CCO was not extended to support minting IAM Roles from inside the cloud-credential-operator.

This does not mean that we will take away Mint mode for the platforms where it exists today.

Reply:
In the absence of #363, the currently recommended pattern is for information about 4.y to be back ported to 4.(y-1) operators, so they can intelligently say "ah, the current cluster state is not compatible with the information I've been taught about 4.y" and set `Upgradeable=False`.  I'm not in favor about moving that responsibility into the CVO.  For example, under some circumstances, the cred operator might be able to list permissions attached to existing credentials, see that they met the expanded 4.y criteria, and set `Upgradeable=True`.  The CVO will never be able to perform that in-cluster cred permissions listing.  I do really appreciate the just-works nature of mint and passthrough mode.  Currently manual mode is a very small fraction of Telemetry-submitting clusters.  But I also understand that handing clusters admin-level creds is a big ask.  If cluster admins want to manually create expanded creds as needed (and ideally reduce perms if we reduce our asks), I am completely fine with that.  But I would prefer the cred-operator have [read-only perms sufficient to audit the creds vs. the cred-requests][1], regardless of who is granting the creds.  I don't see any reason why cluster admins would be uncomfortable with such a read-only cred, but if and only if they are, the cred operator could fall back to the `upgradeable-to` flow, with `Upgradeable=True` and a reason like `BlindManualMode` with a message like `I have no audit creds, so whether you have credentials in place for 4.(y+1) is entirely on you, cluster admins`.  Even then, the cred operator could go `Upgradeable=False` if a cred request it knew 4.(y+1) would ask for had no associated Secret at all.

[1]: https://github.com/openshift/cloud-credential-operator/blame/517f71efb6c5babddc7207b5e6331f51316f5880/README.md#L148

Reply:
For cases where the cred-operator lacks a sufficient cred to audit CredRequest Secrets, it would also be good to have CI around each CredRequest to show that the consuming operator complains clearly if the Secret it has is insufficient to perform a task.  That would make life easier on admins in the completely manual mode to identify and recover if/when they flub a Secret.